libgriddyn

mcneish1 opened the pull request:
This PR adds a new C API to griddyn.

(ignore the first 2 commits, they're identical to those in the units rework PR).

TODO:
 - [x] I'm currently not actually loading model files
 - [x] Nothing's been tested
 - [x] There's no installation target for this
 - [x] The name is currently the same as the library built in src/griddyn; that one ought to be renamed
 - [x] Misc TODOs scattered throughout the implementation
 - [x] Object caching cleanup*
 - [ ] add_test isn't actually adding tests

EDIT 5/1/19:
 - [ ] Generator/Load access should use the actual griddyn interface, rather than the current double indirection + map lookup. It should also have a better name
 - [ ] Generator/Load status should be set with reconnect/disconnect, rather than enable/disable
 - [ ] Since fixed shunt/load determination isn't actually possible with load conductance/susceptance, we don't need them

Notable features:
 - Access every line, bus, load, and generator
   - Note that many features still require C++
 - Access graph topology
 - Disambiguate loads from shunts (manually)
 - Disambiguate generic lines from transformers
 - Enable/disable loads, lines, generators
 - Clear places for new features to go; straightforward implementations of said new features**

\* This should be in the inline TODOs, but it's easily the largest of them
** I hope

mcneish1 committed:
Fix all the rest of the warnings

mcneish1 committed:
Correct indentation of buildSundials.cmake

mcneish1 committed:
Add initial griddyn.h

mcneish1 committed:
Move types, constants to new headers

mcneish1 committed:
Add libgriddyn

phlptp commented on src/CMakeLists.txt:
```diff
@@ -103,6 +103,13 @@ if(BUILD_SHARED_LIBS)
     add_subdirectory(griddyn_shared)
 endif(BUILD_SHARED_LIBS)
 
+# ------------------------------------------------------------
```
Are we anticipating that there will exist both versions of the C library for a little while at least?

mcneish1 replied:
I don't know. Certainly, this one doesn't do everything the other one does, but I wasn't able to find anyone actually using the other one.

phlptp commented on src/libgriddyn/griddyn/bits/griddyn_structs.h:
```diff
@@ -0,0 +1,18 @@
+struct griddyn_bus_ { char _; };
+struct griddyn_link_ { char _; };
+struct griddyn_load_ { char _; };
+struct griddyn_generator_ { char _; };
```
What exactly are you doing here?

mcneish1 replied:
The default struct size for C is zero, but for C++ is one. This standardizes them in case anyone actually creates these structs on the C side for use on the C++ side. 

Unfortunately, they do add size to the C++ objects that inherit from them, because EBO no longer happens. I think an alternative option is just declaring them on the C side, and defining them to be empty on the C++ side? I need to try that, anyway.

mcneish1 replied:
Fixed now, they're declared in C and defined in C++. There shouldn't be name-mangling issues, since everything that uses these structs uses them as pointers, from an `extern "C"` context.

phlptp commented on src/libgriddyn/griddyn/bits/griddyn_types.h:
```diff
@@ -0,0 +1,26 @@
+typedef struct griddyn_link_  griddyn_link;
+typedef struct griddyn_load_  griddyn_load;
+typedef struct griddyn_generator_ griddyn_generator;
+
```
area and relay would also seem to relevent at this level?

mcneish1 replied:
Relay, sure, but none of the models I've seen actually have relays in them. Area's functionality doesn't seem very useful, considering how much of a headache another layer of mapping (like generators/loads, except backwards I think?) it would be.

mcneish1 committed:
Fix griddyn_sim_new to actually load model files

mcneish1 committed:
Fix flag list to refer to strings griddyn looks for, rather than constants that can be set

mcneish1 committed:
Fix -fPIC, library linkage for libgriddyn

mcneish1 committed:
Update libgriddyn with issues noticed in testing
 * griddyn_sim_link_get_adjacency_list was difficult to use sanely,
   since there's no exposed bus index mapping, except implicitly.
   Changed to griddyn_link_get_buses, to enable manual connectivity mapping.
 * misc. functions that require context parameters did not have them in libgriddyn.h
 * vet_pointer usage is now entirely internal, and automatically happens with casts
 * cache coherency is now explicit, with an interface invalidate(). Appropriate calls
   have been placed where necessary.
 * storage access methods have been const-overloaded
 * bus*<->bus index mapping is now part of the general caching behavior
 * sim_interface is now const-correct
 * transformers are additionally detected based on "tap" values != 1

mcneish1 committed:
Add tests for libgriddyn

mcneish1 committed:
Update TODO comments in griddyn_interface

mcneish1 committed:
Add result, status constants to libgriddyn's constants list

mcneish1 committed:
Add installation target for libgriddyn

mcneish1 committed:
Use || instead of or

mcneish1 committed:
Remove 'lib' prefix from internal griddyn library

mcneish1 commented:
Do you know why add_test isn't actually adding tests? I fixed this before, back in January when the units tests were inside GridDyn, but I can't remember what I did.

phlptp commented on config/cmake/buildSundials.cmake:
```diff
@@ -43,41 +43,42 @@ function (build_sundials_msvc)
-    GIT_TAG v3.1.1
-    UPDATE_COMMAND " "
-    BINARY_DIR ${PROJECT_BINARY_DIR}/ThirdParty/sundials
+        cmake_minimum_required(VERSION 3.5)
```
Is this just a reformat?  or was something else changed

mcneish1 replied:
Should just be a reformat

phlptp commented on src/griddyn/Generator.h:
```diff
@@ -27,7 +29,7 @@ class gridSubModel;
 a generator with no dynamics
 */
-class Generator : public gridSecondary
+class Generator : public gridSecondary, public griddyn_generator_
```
I am not sure I like having the multiple inheritance here, especially with a 1 byte size.  That is going to screw with the memory alignment and make for some rather complicated objects.  Is there any way we can Not do this?

mcneish1 replied:
I'll double check that the structs need to be defined on the C-side, rather than just declared. That would let them be empty, at least.

The goal is to return pointers to griddyn things as pointers to real C things in a type safe way. That means no pointers to void. The only way I know of doing this is with a common base object between C/C++, but which isn't common between the C things.

mcneish1 replied:
As they're now empty, the empty base optimization should kick in. Memory alignment should be as before, this should only be visible on the type level.

phlptp commented on src/libgriddyn/src/griddyn_generator_management.cpp:
```diff
@@ -0,0 +1,46 @@
+#include "griddyn_interface.hpp"
+
+griddyn_generator* griddyn_bus_generator_i(griddyn_sim const* ctx, griddyn_bus const* bus, griddyn_idx_t generator_idx)
```
What does the _i  mean?

mcneish1 replied:
It means I couldn't come up with a better name for array access. Do you have any suggestions?
 
(I don't really like this for generators/loads, since it has to go through the bus pointer map. A hypothetical `griddyn_bus_i_generator_i` wouldn't require that, but would also want better naming.)

phlptp commented on src/libgriddyn/src/griddyn_run.cpp:
```diff
@@ -0,0 +1,58 @@
+{
+  griddyn::gridDynSimulation* sim_ptr = impl::sim_cast(ctx);
+  {
+    // TODO: There has got to be a better way of doing this
```
I believe there is a set flags function that uses an index, not sure that is public or not

phlptp commented:
How are you doing the export for a shared library I didn't see any EXPORT macros in front of the functions?

mcneish1 commented:
>How are you doing the export for a shared library I didn't see any EXPORT macros in front of the functions?

https://github.com/LLNL/GridDyn/pull/48/files#diff-ab31dc62f7a2318643e42dc2490c6030R14

They all have `LIBGRIDDYN_EXPORT`, defined here: https://github.com/LLNL/GridDyn/pull/48/files#diff-c57c0bcf55a58dbebe8869af57db1836

mcneish1 committed:
Reduce size of the C compatibility structs

phlptp commented:
@mcneish1   What is the status of this,  are you still working on it or is it ready for final review and merge?

mcneish1 commented:
I should have time to fix the rest of my issues and write the paragraph on this later this week.
